Added git_config_find_local()

swisspol opened the pull request.

swisspol commented:
There are APIs to find the path of other config files, but not the local one. Even if it's trivial, it's nice to have libgit2 provide a way to find this path for consistency.

ethomson commented on src/config.c:
```diff
@@ -1067,6 +1067,22 @@ void git_config_iterator_free(git_config_iterator *iter)
 	iter->free(iter);
 }
 
+int git_config_find_local(git_repository *repo, git_buf *path)
```
Our "out" params are first in the decl, this would be better as:

    int git_config_find_local(git_buf *path, git_repository *repo);

ethomson commented on src/config.c:
```diff
@@ -1067,6 +1067,22 @@ void git_config_iterator_free(git_config_iterator *iter)
+
+	git_buf_sanitize(path);
+	if (git_buf_joinpath(path, repo->path_repository, GIT_CONFIG_FILENAME_INREPO) < 0)
+    return -1;
```
Minor spaces vs tabs formatting nit here.

ethomson commented:
Makes sense.  Does it get the submodule git-dir redirection correct?  I would imagine so but I would love to see a quick unit test.

swisspol commented:
I addressed the comments above.

> Does it get the submodule git-dir redirection correct?

I'm not sure I understand what could be the problem here? This API uses the exact same approach as `load_config()` which I assume is based on `repo->path_repository` being valid in the first place, submodule or not. Essentially adding a unit test for this would be testing something that `git_repository_config__weakptr()` doesn't even test AFAIK.

How about instead slightly changing `git_repository_config__weakptr()` so that it uses this new API like it uses `git_config_find_global()` and friends? This way, it's the same code path all around (and presumably very well tested).

swisspol commented:
The build failed in AppVeyor but not because of the PR it seems.

nulltoken commented:
Restarted it.

Interestingly I regularly stumble against this one as well with LibGit2Sharp builds. BitBucket issue?

```
Failure:
online::clone::bitbucket_style [c:/projects/libgit2/tests/online/clone.c:313]
Function call failed: (git_clone(&g_repo, "https://libgit3:wrong@bitbucket.org/libgit2/testgitrepository.git", "./foo", &g_options))
   error -1 - failed to send request: The buffers supplied to a function was too small.
```

carlosmn commented:
>Interestingly I regularly stumble against this one as well with LibGit2Sharp builds. BitBucket issue?

It looks like at times WinHTTP wants to give us more data than we let it, though that doesn't make a lot of sense. It might also just be an old error when we were trying to find out the size of something, and then we never clear the error.

swisspol commented:
So @ethomson what should I do about the above?

carlosmn commented:
The nature of this function is rather different from the other ones. The `_find_` in the other functions comes from the fact that they need to go out and figure out which is the most appropriate file to use for that purpose.

We know exactly where the repository's configuration file is, we need to go find it. This feels a lot more like a method on a repository. But since you already have a repository, you can grab the configuration object directly, so I'm not sure where this is expected to be more useful than using the object that's already there to be used.

swisspol commented:
Even if you know where the local config file is located, you still need a piece of code to find / build that path. Right now, it's built into `load_config()` instead of being isolated in a separate function like for the other config files. These functions are exposed as APIs to clients presumably because they might need to do something with the paths of config files. Then why not expose an API for the local config file path as well?

Would changing the API name address your concerns?

swisspol commented:
Any final feedback on this? Should I abandon this PR?

carlosmn commented:
The reasons why this function doesn't fit in this namespace haven't changed. It's not independent of any repository and it's not something we have to find.

The subject here is the repository, so it should be a method there. You also mention that this extracts the logic which the repository currently uses to load the config, but this PR doesn't change that, so this is duplicating the logic, which goes against the reason for exposing such a function. If this is meant to extract that logic, then the old code should be replaced by a call to this method.

ethomson commented:
So yeah, I agree that *if* we're going to make this public API, then this should be `git_repository_config_path`.  And the existing `git_repository` code should use this, and it shouldn't be duplicated code.  Otherwise, it's fairly simple to compose this on your own.

carlosmn commented:
Closing for inactivity.
